File: clients/python/src/graphdb_client/_transport.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Mapping

import httpx

from ._retry import RetryConfig, compute_delay, is_retryable, parse_retry_after
from .errors import from_response


@dataclass
class ApiResult:
    data: Any
    headers: Mapping[str, str]


class Transport:
    """Single choke point for every HTTP request: auth, error mapping, raw access."""

    def __init__(
        self,
        base_url: str,
        *,
        token: str | None = None,
        api_key: str | None = None,
        username: str | None = None,
        password: str | None = None,
        timeout: float = 30.0,
        retries: RetryConfig | None = None,
    ) -> None:
        if (username is None) != (password is None):
            raise ValueError("username and password must be provided together")
        self._token = token
        self._api_key = api_key
        self._username = username
        self._password = password
        self._refresh_token: str | None = None
        self._http = httpx.Client(base_url=base_url.rstrip("/"), timeout=timeout)
        self._retries = retries if retries is not None else RetryConfig()

    def _auth_headers(self) -> dict[str, str]:
        return build_auth_headers(self._token, self._api_key)

    def _has_credentials(self) -> bool:
        return self._username is not None and self._password is not None

    def _login(self) -> None:
        resp = self._http.post(
            "/auth/login", json={"username": self._username, "password": self._password}
        )
        if resp.status_code >= 400:
            raise from_response(resp.status_code, _safe_json(resp), "POST", "/auth/login")
        data = _safe_json(resp) or {}
        self._token = data.get("access_token")
        self._refresh_token = data.get("refresh_token")

    def _refresh(self) -> bool:
        """Return True if the token was refreshed. Falls back to re-login."""
        if self._refresh_token:
            resp = self._http.post("/auth/refresh", json={"refresh_token": self._refresh_token})
            if resp.status_code < 400:
                self._token = (_safe_json(resp) or {}).get("access_token")
                return True
        if self._has_credentials():
            self._login()
            return True
        return False

    def _attempt(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: Mapping[str, Any] | None = None,
    ) -> httpx.Response:
        if self._token is None and self._has_credentials():
            self._login()

        resp = self._http.request(
            method, path, json=json, params=params, headers=self._auth_headers()
        )

        if resp.status_code == 401 and (self._refresh_token or self._has_credentials()):
            if self._refresh():
                resp = self._http.request(
                    method, path, json=json, params=params, headers=self._auth_headers()
                )
        return resp

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: Mapping[str, Any] | None = None,
    ) -> ApiResult:
        attempt = 0
        while True:
            try:
                resp = self._attempt(method, path, json=json, params=params)
            except httpx.TransportError as exc:
                if attempt < self._retries.max_retries and is_retryable(
                    method, None, exc, self._retries
                ):
                    time.sleep(compute_delay(attempt, self._retries, None))
                    attempt += 1
                    continue
                raise

            if (
                resp.status_code >= 400
                and attempt < self._retries.max_retries
                and is_retryable(method, resp.status_code, None, self._retries)
            ):
                retry_after = (
                    parse_retry_after(resp.headers.get("Retry-After"))
                    if self._retries.respect_retry_after
                    else None
                )
                time.sleep(compute_delay(attempt, self._retries, retry_after))
                attempt += 1
                continue

            if resp.status_code >= 400:
                raise from_response(resp.status_code, _safe_json(resp), method, path)
            return ApiResult(data=_safe_json(resp), headers=resp.headers)

    def close(self) -> None:
        self._http.close()


def build_auth_headers(token: str | None, api_key: str | None) -> dict[str, str]:
    headers: dict[str, str] = {}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    elif api_key:
        headers["X-API-Key"] = api_key
    return headers


def _safe_json(resp: httpx.Response) -> Any:
    if not resp.content:
        return None
    try:
        return resp.json()
    except ValueError:
        return resp.text

File: clients/go/transport.go
package graphdb

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

type apiResult struct {
	status int
	data   json.RawMessage
	header http.Header
}

type transport struct {
	baseURL string
	http    *http.Client

	// immutable after New
	apiKey   string
	username string
	password string

	// mu guards the mutable token fields (refresh rewrites them while other
	// goroutines read them for auth headers).
	mu           sync.Mutex
	token        string
	refreshToken string

	// refreshMu serializes login/refresh so concurrent 401s coalesce into a
	// single token exchange. Never held while mu is held.
	refreshMu sync.Mutex

	maxRetries int
}

func (t *transport) currentToken() string {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.token
}

func (t *transport) authHeaders(h http.Header, token string) {
	switch {
	case token != "":
		h.Set("Authorization", "Bearer "+token)
	case t.apiKey != "":
		h.Set("X-API-Key", t.apiKey)
	}
}

// request performs a JSON request with retry on retryable statuses. On a first
// 401 it refreshes (or lazily logs in) and retries once.
func (t *transport) request(ctx context.Context, method, path string, body any, params url.Values) (*apiResult, error) {
	if t.username != "" {
		if err := t.loginIfNeeded(ctx); err != nil {
			return nil, err
		}
	}
	refreshed := false
	retries := 0
	for {
		// Snapshot the token used for this attempt so the 401 handler can
		// tell whether a concurrent caller already replaced it.
		usedToken := t.currentToken()
		resp, err := t.attempt(ctx, method, path, body, params, usedToken)
		if err != nil {
			return nil, err
		}
		if resp.StatusCode == http.StatusUnauthorized && !refreshed && t.usesLogin() {
			resp.Body.Close()
			refreshed = true
			if err := t.refreshIfStale(ctx, usedToken); err != nil {
				return nil, err
			}
			continue // retry once with the new token; does NOT consume the retry budget
		}
		if resp.StatusCode >= 400 && retries < t.maxRetries && isRetryable(resp.StatusCode) {
			resp.Body.Close()
			select {
			case <-time.After(backoff(retries)):
			case <-ctx.Done():
				return nil, ctx.Err()
			}
			retries++
			continue
		}
		defer resp.Body.Close()
		data, readErr := io.ReadAll(resp.Body)
		if readErr != nil {
			return nil, readErr
		}
		if resp.StatusCode >= 400 {
			return nil, fromResponse(resp.StatusCode, data, method, path)
		}
		return &apiResult{status: resp.StatusCode, data: data, header: resp.Header}, nil
	}
}

// rawAttempt performs a single unauthenticated request and returns the raw
// body. It is used only by login/refresh: no Authorization header is sent
// (credentials travel in the body, and a stale Bearer would leak to
// auth-adjacent logs or be rejected by gateways that validate it globally),
// and it must not re-enter the 401-refresh path in request().
func (t *transport) rawAttempt(ctx context.Context, method, path string, body any) (json.RawMessage, error) {
	resp, err := t.attempt(ctx, method, path, body, nil, "")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, readErr := io.ReadAll(resp.Body)
	if readErr != nil {
		return nil, readErr
	}
	if resp.StatusCode >= 400 {
		return nil, fromResponse(resp.StatusCode, data, method, path)
	}
	return data, nil
}

func (t *transport) attempt(ctx context.Context, method, path string, body any, params url.Values, token string) (*http.Response, error) {
	var rdr io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		rdr = bytes.NewReader(b)
	}
	u := strings.TrimRight(t.baseURL, "/") + path
	if len(params) > 0 {
		u += "?" + params.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, method, u, rdr)
	if err != nil {
		return nil, err
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	t.authHeaders(req.Header, token)
	return t.http.Do(req)
}

func isRetryable(status int) bool {
	return status == http.StatusTooManyRequests || status >= 500
}

func backoff(attempt int) time.Duration {
	// 100<<attempt overflows for large attempt values; the cap is reached at
	// attempt 5 anyway.
	if attempt > 4 {
		return 2 * time.Second
	}
	d := time.Duration(100<<attempt) * time.Millisecond
	if d > 2*time.Second {
		d = 2 * time.Second
	}
	return d
}
